wa/workloads/recentfling/__init__.py

```diff
@@ -0,0 +1,126 @@
+                  If set, recentfling will use the fling parameters for this
+                  device instead of automatically guessing the device.  This can
+                  also be used if the device is not supported by recentfling,
+                  but its screensize is similar to that of one that is supported.
+
+                  For possible values, check your recentfling.sh.  At the time
+                  of writing, valid values are: 'shamu', 'hammerhead', 'angler',
+                  'ariel', 'mtp8996', 'bullhead' or 'volantis'.
+                  """),
+    ]
+
+    def initialise(self, context):  # pylint: disable=no-self-use
+        if context.target.get_sdk_version() < 23:
+            raise WorkloadError("This workload relies on ``dumpsys gfxinfo`` \
+                                 only present in Android M and onwards")
+
+    def setup(self, context):
+        self.defs_host = context.resolver.get(File(self, "defs.sh"))
+        self.defs_target = self.target.install(self.defs_host)
+        self.recentfling_host = context.resolver.get(File(self, "recentfling.sh"))
+        self.recentfling_target = self.target.install(self.recentfling_host)
+        self._kill_recentfling()
+        self.target.ensure_screen_is_on()
+
+    def run(self, context):
+        args = '-i {} '.format(self.loops)
+        if not self.start_apps:
+            args += '-N '
+        if self.device_name is not None:
+            args += '-d {}'.format(self.device_name)
+        cmd = "echo $$>{dir}/pidfile; cd {bindir}; exec ./recentfling.sh {args}; rm {dir}/pidfile"
+        cmd = cmd.format(args=args, dir=self.target.working_directory,
+                         bindir=self.target.executables_directory)
+        try:
+            self.output = self.target.execute(cmd, timeout=120)
+        except KeyboardInterrupt:
+            self._kill_recentfling()
+            raise
+
+    def update_result(self, context):
+        group_names = ["90th Percentile", "95th Percentile", "99th Percentile", "Jank", "Jank%"]
+        count = 0
+        for line in self.output.strip().splitlines():
+            p = re.compile("Frames: \d+ latency: (?P<pct90>\d+)/(?P<pct95>\d+)/(?P<pct99>\d+) Janks: (?P<jank>\d+)\((?P<jank_pct>\d+)%\)")
+            match = p.search(line)
+            if match:
+                count += 1
+                if line.startswith("AVE: "):
+                    group_names = ["Average " + g for g in group_names]
+                    count = 0
+                for metric in zip(group_names, match.groups()):
+                    context.result.add_metric(metric[0],
+                                              metric[1],
+                                              None,
+                                              classifiers={"loop": count or "Average"})
+
+    def teardown(self, context):
+        self.target.uninstall_executable(self.recentfling_target)
+        self.target.uninstall_executable(self.defs_target)
```

Comment:
If installation is moved into `initialize`, as suggested above, this should correspondingly be done inside `finalize`.